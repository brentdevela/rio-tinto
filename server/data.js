const ideas = [{
    id: '1',
    createdDate: '2020-02-01',
    title: 'My Tile',
    body: 'This is my idea!'
},{
    id: '2',
    createdDate: '2020-02-02',
    title: 'Your Tile',
    body: 'This was your idea!'
},{
    id: '3',
    createdDate: '2020-02-03',
    title: 'Their Tile',
    body: 'This was their idea!'
},{
    id: '4',
    createdDate: '2020-02-04',
    title: 'Silly Tile',
    body: 'What a silly idea!'
},{
    id: '5',
    createdDate: '2020-02-05',
    title: 'Boring Tile',
    body: 'No idea!'
},{
    id: '6',
    createdDate: '2020-02-06',
    title: 'Happy Tile',
    body: 'Great idea!!'
},{
    id: '7',
    createdDate: '2020-02-07',
    title: 'Sad Tile',
    body: 'That was a conservative idea!'
},{
    id: '8',
    createdDate: '2020-02-08',
    title: 'Angry Tile',
    body: 'What a liberal idea!'
},{
    id: '9',
    createdDate: '2020-02-09',
    title: 'Progressive Tile',
    body: 'Isnt that a progressive idea!'
},{
    id: '10',
    createdDate: '2020-02-10',
    title: 'Crazy Tile',
    body: 'What a liberal idea!'
},{
    id: '11',
    createdDate: '2020-02-11',
    title: 'Indifferent Tile',
    body: 'What an indifferent idea!'
}];

const newIdea = {
    id: '',
    createdDate: '',
    title: '',
    body: ''
};

module.exports = {
    ideas,
    newIdea
};

