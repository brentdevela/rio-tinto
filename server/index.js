'use strict';
const port = 8010;
const server = require('./server')();
server.listen(port, () => console.log(`App started and listening on port ${port}`));