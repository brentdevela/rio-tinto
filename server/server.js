'use strict';
const lodash = require('lodash');
const moment = require('moment');
const express = require('express');
const cors = require('cors');
const app = express();

app.use(cors());

const bodyParser = require('body-parser');
const jsonParser = bodyParser.json();

const data = require('./data');

module.exports = () => {
    app.get('/ideas', (req, res) => {
        res.send(data.ideas); // DUMMY
    });

    app.post('/idea', jsonParser, (req, res) => {
        const clonedIdea = lodash.cloneDeep(data.newIdea);
        clonedIdea.id = String(Number(lodash.last(data.ideas).id) + 1);
        clonedIdea.createdDate = moment().format('YYYY-MM-DD');
        data.ideas.push(clonedIdea);
        res.send(clonedIdea);
    });

    app.put('/idea', jsonParser, (req, res) => {
        const idea = req.body;
        const index = lodash.findIndex(data.ideas, (currentIdea) => currentIdea.id === idea.id);
        data.ideas[index] = idea;
        res.send(req.body);
    });

    app.del('/idea/:id', jsonParser, (req, res) => {
        lodash.remove(data.ideas, (idea) => idea.id === req.params.id);
        res.send(req.body);
    });

    return app;
};
