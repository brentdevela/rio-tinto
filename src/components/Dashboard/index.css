.dashboard {
    max-width: 1210px;
}

.add-idea {
    padding: 10px;
    text-align: left;
}