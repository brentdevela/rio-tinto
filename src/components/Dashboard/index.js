import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { Container, Row, Col, Button } from 'react-bootstrap';
import { IoIosAddCircleOutline } from 'react-icons/io';
import Idea from '../../models/Idea';
import Tile from '../Tile';
import { noop } from 'lodash';

import './index.css';

const Dashboard = (props) => {

    const [ newIdea, setNewIdea ] = useState(false);

    useEffect(() => {
        if (props.ideas.length === 0) {
            props.onLoadIdeas();
        }
    });

    const onClickNewIdea = () => {
        props.onNewIdea()
            .then(() => {
                setNewIdea(true);
            });
    };

    const onChangeIdea = (idea) => {
      props.onChangeIdea(idea)
          .then(() => {
              setNewIdea(false);
          }) ;
    };

    const onDeleteIdea = (idea) => {
        props.onDeleteIdea(idea)
            .then(() => {
                setNewIdea(false);
            }) ;
    }

    return (
        <Container fluid className="dashboard">
            <Row>
                <Col>
                    <div className="add-idea d-flex flex-column">
                        <Button variant="primary" onClick={onClickNewIdea}>
                            <IoIosAddCircleOutline /> New Idea
                        </Button>
                    </div>
                </Col>
            </Row>
            <Row>
                {props.ideas.map((idea, idx) =>
                    (<Col key={idea.id} xs>
                        <Tile
                            idea={idea}
                            onChangeIdea={onChangeIdea}
                            onDeleteIdea={onDeleteIdea}
                            isFocused={newIdea && idx === (props.ideas.length - 1)}
                        />
                    </Col>))}
            </Row>
        </Container>
    )
};

Dashboard.propTypes = {
    ideas: PropTypes.arrayOf(Idea).isRequired,
    onLoadIdeas: PropTypes.func,
    onNewIdea: PropTypes.func,
    onChangeIdea: PropTypes.func,
    onDeleteIdea: PropTypes.func,
};

Dashboard.defaultProps = {
  onNewIdea: noop
};


export default Dashboard;