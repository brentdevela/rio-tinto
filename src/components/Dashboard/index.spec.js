import React from 'react';
import { shallow } from 'enzyme';
import { noop } from 'lodash';
import Dashboard from './index';

import { ideas } from '../../../server/data';

describe('<Dashboard>', () => {
    const render = (props) => {
        return shallow(
          <Dashboard
              {...props}
          />
        );
    };

    it('should render Dashboard', () => {
        const dashboard = render({ ideas });
        expect(dashboard).toMatchSnapshot();
    });

    it('should trigger onNewIdea', () => {
        const onNewIdea = jest.fn().mockReturnValue(Promise.resolve(true));
        const dashboard = render({ ideas, onNewIdea });
        const newIdeaButton = dashboard.find('Button').first();
        newIdeaButton.simulate('click');
        expect(onNewIdea).toBeCalled();
    });

    it('should trigger onChangeIdea', () => {
        const onChangeIdea = jest.fn().mockReturnValue(Promise.resolve(true));
        const dashboard = render({ ideas, onChangeIdea });
        const tile = dashboard.find('Tile').first();
        tile.prop('onChangeIdea')(ideas[0]);
        expect(onChangeIdea).toBeCalled();
    });

    it('should trigger onDeleteIdea', () => {
        const onDeleteIdea = jest.fn().mockReturnValue(Promise.resolve(true));
        const dashboard = render({ ideas, onDeleteIdea });
        const tile = dashboard.find('Tile').first();
        tile.prop('onDeleteIdea')(ideas[0]);
        expect(onDeleteIdea).toBeCalled();
    });
});