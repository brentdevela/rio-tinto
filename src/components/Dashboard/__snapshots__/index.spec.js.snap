// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard> should render Dashboard 1`] = `ShallowWrapper {}`;
